src/core/service.c

```diff
@@ -2758,6 +2760,65 @@ static void service_run_next_main(Service *s) {
         (void) service_set_main_pidref(s, TAKE_PIDREF(pidref), &s->main_command->exec_status.start_timestamp);
 }
 
+static bool service_needs_reload_extensions(Unit *u) {
+        Service *s = ASSERT_PTR(SERVICE(u));
+
+        /* Only support this for notify-reload service types. */
+        if (s->type != SERVICE_NOTIFY_RELOAD)
+                return false;
+
+        /* For now, opportunistically reload as long as we have an extension,
+         * and do not try to actually figure out if we have a vpick'd extension
+         * or not. */
+        return s->exec_context.n_extension_images > 0 ||
+                        !strv_isempty(s->exec_context.extension_directories);
+}
+
+static int service_reload_extensions(Unit *u) {
+        /* TODO: do this asynchronously */
+        _cleanup_free_ char *propagate_dir = NULL, *private_namespace_dir = NULL, *incoming_dir = NULL;
+
+        Service *s = ASSERT_PTR(SERVICE(u));
+
+        /* TODO: Add support for user services, which can use
+         * ExtensionDirectories= + notify-reload. For now, skip for user
+         * services. */
+        if (u->manager->runtime_scope != RUNTIME_SCOPE_SYSTEM) {
+                log_unit_info(UNIT(s), "Extension refresh is not currently supported for user service.");
+                return 0;
+        }
+
+        private_namespace_dir = strdup("/run/systemd");
+        if (!private_namespace_dir)
+                return -ENOMEM;
+
+        propagate_dir = path_join("/run/systemd/propagate/", u->id);
+        if (!propagate_dir)
+                return -ENOMEM;
+
+        incoming_dir = strdup("/run/systemd/incoming");
```

Comment:
It is not necessary to strdup this and the one above, as they are fixed strings, just initialize inline below to the raw values